packages/visibility_detector/lib/src/render_visibility_detector.dart

```diff
@@ -34,33 +98,220 @@ class RenderVisibilityDetector extends RenderProxyBox {
 
   /// Used by [VisibilityDetector.updateRenderObject].
   set onVisibilityChanged(VisibilityChangedCallback? value) {
+    if (_onVisibilityChanged == value) {
+      return;
+    }
+    _compositionCallbackCanceller?.call();
+    _compositionCallbackCanceller = null;
     _onVisibilityChanged = value;
-    markNeedsCompositingBitsUpdate();
-    markNeedsPaint();
+
+    if (value == null) {
+      // Remove all cached data so that we won't fire visibility callbacks when
+      // a timer expires or get stale old information the next time around.
+      forget(key);
+      _lastVisibility.remove(key);
+    } else {
+      markNeedsPaint();
+      // If an update is happening and some ancestor no longer paints this RO,
+      // the markNeedsPaint above will never cause the composition callback to
+      // fire and we could miss a hide event. This schedule will get
+      // over-written by subsequent updates in paint, if paint is called.
+      _scheduleUpdate(null);
```

Comment:
nit: maybe make the parameters a required optional one so this call side reads a little better...

Reply:
Going to make it an optional optional :)